consensus/byzantine_test.go

```diff
@@ -349,6 +349,230 @@ func TestByzantineConflictingProposalsWithPartition(t *testing.T) {
+
+				// parts
+				for i := 0; i < int(blockParts1.Total()); i++ {
+					part := blockParts1.GetPart(i)
+					msg := &BlockPartMessage{
+						Height: height,
+						Round:  round,
+						Part:   part,
+					}
+					peer.Send(DataChannel, MustEncode(msg))
+				}
+
+				if i == VICTIM_PEER_INDEX {
+					peer.Send(VoteChannel, MustEncode(&VoteMessage{prevote}))
+					peer.Send(VoteChannel, MustEncode(&VoteMessage{precommit}))
+				} else {
+					peer.Send(VoteChannel, MustEncode(&VoteMessage{prevote}))
+					peer.Send(VoteChannel, MustEncode(&VoteMessage{precommitNil}))
+				}
+			}
+		} else if height == 1 && round == 1 {
+			// We sign anything our friends need!
+			prevote, _ := cs.signVote(tmproto.PrevoteType, cs.ProposalBlock.Hash(), cs.ProposalBlockParts.Header())
+			precommit, _ := cs.signVote(tmproto.PrecommitType, cs.ProposalBlock.Hash(), cs.ProposalBlockParts.Header())
+
+			for i, peer := range peers {
+				if i != VICTIM_PEER_INDEX {
+					peer.Send(VoteChannel, MustEncode(&VoteMessage{prevote}))
+					peer.Send(VoteChannel, MustEncode(&VoteMessage{precommit}))
+				}
+			}
+		}
+	}
+
+	css[BYZ_INDEX].decideProposal = func(height int64, round int32) {}
+	css[BYZ_INDEX].setProposal = func(proposal *types.Proposal) error {
+		if proposal.Height == 1 && proposal.Round == 0 {
+			// this is us. Let's fork the tendermint
+		} else if proposal.Height == 1 && proposal.Round == 1 {
+			// Lets help friends to make a fork
+			return css[BYZ_INDEX].defaultSetProposal(proposal)
+		} else if proposal.Height == 2 && proposal.Round == 0 {
+			// Already forked!
+		}
+		return nil
+	}
+	// ----------------------------------------------------------------------
+
+	defer func() {
+		for _, sw := range switches {
+			err := sw.Stop()
+			require.NoError(t, err)
+		}
+	}()
+
+	// start the non-byz state machines.
+	// note these must be started before the byz
+	for i := 0; i < N; i++ {
+		if i != BYZ_INDEX {
+			cr := reactors[i].(*Reactor)
+			cr.SwitchToConsensus(cr.conS.GetState(), false)
+		}
+	}
+
+	// start the byzantine state machine
+	byzR := reactors[BYZ_INDEX].(*ByzantineReactor)
+	s := byzR.reactor.conS.GetState()
+	byzR.reactor.SwitchToConsensus(s, false)
+
+	// byz proposer sends one block to peers[0]
+	// and the other block to peers[1] and peers[2].
+	// note peers and switches order don't match.
+	peers := switches[BYZ_INDEX].Peers().List()
+
+	// partition A
+	ind0 := getSwitchIndex(switches, peers[VICTIM_PEER_INDEX])
+
+	// partition B
+	ind1 := getSwitchIndex(switches, peers[0])
+	ind2 := getSwitchIndex(switches, peers[2])
+	p2p.Connect2Switches(switches, ind1, ind2)
+
+	m := 0
+	for {
+		for i := 0; i < N; i++ {
+			v := css[BYZ_INDEX].Votes.Precommits(0).GetByIndex(int32(i))
+			if v != nil {
+				if v.BlockID.Hash != nil {
+					peers[VICTIM_PEER_INDEX].Send(VoteChannel, MustEncode(&VoteMessage{v}))
+					m |= (0x1 << i)
+				}
+			}
+		}
+		time.Sleep(10 * time.Millisecond)
+		if m == 0x6 {
+			break
+		}
+	}
```

Comment:
Can you explain to me what this for loop is doing? It doesn't seem like the test ever leaves this for loop (i.e. I never see the log "A block has been committed. Healing partition" in the output)